src/pybind/mgr/dashboard/services/cephfs.py

```diff
@@ -45,10 +45,6 @@ def __init__(self, fs_name=None):
             self.cfs.mount()
         logger.debug("mounted cephfs filesystem")
 
-    def __del__(self):
-        logger.debug("shutting down cephfs filesystem")
-        self.cfs.shutdown()
```

Comment:
It looks like this is removed without a replacement?  Probably need a `with` context or something like that around `_cephfs_instance()` in `src/pybind/mgr/dashboard/controllers/cephfs.py` to compensate.

Reply:
@wido Did you resolve this by mistake?  Not seeing anything in your approval message related to the concern that I raised here.

Reply:
Oh yes, that was by mistake!

Reply:
> It looks like this is removed without a replacement? Probably need a `with` context or something like that around `_cephfs_instance()` in `src/pybind/mgr/dashboard/controllers/cephfs.py` to compensate.

Maybe cython libcephfs's LibCephFS.`__dealloc__()` compensates for this, https://github.com/ceph/ceph/blob/v18.2.0/src/pybind/cephfs/cephfs.pyx#L573 . It calls LibCephFS.shutdown().

https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#finalization-methods-dealloc-and-del

If the above is not sufficient, I will continue working on the following **untested** changes that adds `__enter__() `and `__exit__()` to mgr/dashboard/services/cephfs.py's CephFS similar to those in  pybind/cephfs/cephfs.pyx's LibCephFS, https://github.com/ceph/ceph/blob/v18.2.0/src/pybind/cephfs/cephfs.pyx#L565

<pre>
diff --git a/src/pybind/mgr/dashboard/services/cephfs.py b/src/pybind/mgr/dashboard/services/cephfs.py
index ffbf9d0c816..a1156cb9751 100644
--- a/src/pybind/mgr/dashboard/services/cephfs.py
+++ b/src/pybind/mgr/dashboard/services/cephfs.py
@@ -38,13 +38,19 @@ class CephFS(object):
     def __init__(self, fs_name=None):
         logger.debug("initializing cephfs connection")
         self.cfs = cephfs.LibCephFS(rados_inst=mgr.rados)
-        logger.debug("mounting cephfs filesystem: %s", fs_name)
-        if fs_name:
-            self.cfs.mount(filesystem_name=fs_name)
+        self.fs_name = fs_name
+
+    def __enter__(self):
+        logger.debug("mounting cephfs filesystem: %s", self.fs_name)
+        if self.fs_name:
+            self.cfs.mount(filesystem_name=self.fs_name)
         else:
             self.cfs.mount()
         logger.debug("mounted cephfs filesystem")
 
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.cfs.shutdown()
+
     @contextmanager
     def opendir(self, dirpath):
         d = None
diff --git a/src/pybind/mgr/dashboard/controllers/cephfs.py b/src/pybind/mgr/dashboard/controllers/cephfs.py
index ed83f91d0c9..0d4624fc235 100644
--- a/src/pybind/mgr/dashboard/controllers/cephfs.py
+++ b/src/pybind/mgr/dashboard/controllers/cephfs.py
@@ -418,7 +418,9 @@ class CephFS(RESTController):
         :rtype: dict
         """
         try:
-            return self._get_root_directory(self._cephfs_instance(fs_id))
+            cfs = self._cephfs_instance(fs_id)
+            with cfs:
+                return self._get_root_directory(cfs)
         except (cephfs.PermissionError, cephfs.ObjectNotFound):  # pragma: no cover
             return None
 
@@ -450,7 +452,8 @@ class CephFS(RESTController):
         path = self._set_ls_dir_path(path)
         try:
             cfs = self._cephfs_instance(fs_id)
-            paths = cfs.ls_dir(path, depth)
+            with cfs:
+                paths = cfs.ls_dir(path, depth)
         except (cephfs.PermissionError, cephfs.ObjectNotFound):  # pragma: no cover
             paths = []
         return paths
@@ -479,7 +482,8 @@ class CephFS(RESTController):
         :param path: The path of the directory.
         """
         cfs = self._cephfs_instance(fs_id)
-        cfs.mk_dirs(path)
+        with cfs:
+            cfs.mk_dirs(path)
 
     @RESTController.Resource('DELETE', path='/tree')
     def rm_tree(self, fs_id, path):
@@ -489,7 +493,8 @@ class CephFS(RESTController):
         :param path: The path of the directory.
         """
         cfs = self._cephfs_instance(fs_id)
-        cfs.rm_dir(path)
+        with cfs:
+            cfs.rm_dir(path)
 
     @RESTController.Resource('PUT', path='/quota')
     @allow_empty_body
@@ -502,7 +507,8 @@ class CephFS(RESTController):
         :param max_files: The file limit.
         """
         cfs = self._cephfs_instance(fs_id)
-        return cfs.set_quotas(path, max_bytes, max_files)
+        with cfs:
+            return cfs.set_quotas(path, max_bytes, max_files)
 
     @RESTController.Resource('GET', path='/quota')
     @EndpointDoc("Get Cephfs Quotas of the specified path",
@@ -521,7 +527,8 @@ class CephFS(RESTController):
         :rtype: dict
         """
         cfs = self._cephfs_instance(fs_id)
-        return cfs.get_quotas(path)
+        with cfs:
+            return cfs.get_quotas(path)
 
     @RESTController.Resource('POST', path='/write_to_file')
     @allow_empty_body
@@ -533,7 +540,8 @@ class CephFS(RESTController):
         :param buf: The str to write to the buf.
         """
         cfs = self._cephfs_instance(fs_id)
-        cfs.write_to_file(path, buf)
+        with cfs:
+            cfs.write_to_file(path, buf)
 
     @RESTController.Resource('DELETE', path='/unlink')
     def unlink(self, fs_id, path) -> None:
@@ -543,7 +551,8 @@ class CephFS(RESTController):
         :param path: The path of the file or link to unlink.
         """
         cfs = self._cephfs_instance(fs_id)
-        cfs.unlink(path)
+        with cfs:
+            cfs.unlink(path)
 
     @RESTController.Resource('GET', path='/statfs')
     @EndpointDoc("Get Cephfs statfs of the specified path",
@@ -562,7 +571,8 @@ class CephFS(RESTController):
         :rtype: dict
         """
         cfs = self._cephfs_instance(fs_id)
-        return cfs.statfs(path)
+        with cfs:
+            return cfs.statfs(path)
 
     @RESTController.Resource('POST', path='/snapshot')
     @allow_empty_body
@@ -577,15 +587,16 @@ class CephFS(RESTController):
         :rtype: str
         """
         cfs = self._cephfs_instance(fs_id)
-        list_snaps = cfs.ls_snapshots(path)
-        for snap in list_snaps:
-            if name == snap['name']:
-                raise DashboardException(code='Snapshot name already in use',
-                                         msg='Snapshot name {} is already in use.'
-                                         'Please use another name'.format(name),
-                                         component='cephfs')
+        with cfs:
+            list_snaps = cfs.ls_snapshots(path)
+            for snap in list_snaps:
+                if name == snap['name']:
+                    raise DashboardException(code='Snapshot name already in use',
+                                             msg='Snapshot name {} is already in use.'
+                                             'Please use another name'.format(name),
+                                             component='cephfs')
 
-        return cfs.mk_snapshot(path, name)
+            return cfs.mk_snapshot(path, name)
 
     @RESTController.Resource('DELETE', path='/snapshot')
     def rm_snapshot(self, fs_id, path, name):
@@ -596,7 +607,8 @@ class CephFS(RESTController):
         :param name: The name of the snapshot.
         """
         cfs = self._cephfs_instance(fs_id)
-        cfs.rm_snapshot(path, name)
+        with cfs:
+            cfs.rm_snapshot(path, name)
 
 
 class CephFSClients(object):
@@ -661,7 +673,8 @@ class CephFsUi(CephFS):
         path = self._set_ls_dir_path(path)
         try:
             cfs = self._cephfs_instance(fs_id)
-            paths = cfs.ls_dir(path, depth)
+            with cfs:
+                paths = cfs.ls_dir(path, depth)
             if path == os.sep:
                 paths = [self._get_root_directory(cfs)] + paths
         except (cephfs.PermissionError, cephfs.ObjectNotFound):  # pragma: no cover
diff --git a/src/pybind/mgr/dashboard/controllers/nfs.py b/src/pybind/mgr/dashboard/controllers/nfs.py
index 36b88d76b16..687f2e171ce 100644
--- a/src/pybind/mgr/dashboard/controllers/nfs.py
+++ b/src/pybind/mgr/dashboard/controllers/nfs.py
@@ -254,8 +254,9 @@ class NFSGaneshaUi(BaseController):
         try:
             cfs = CephFS(fs_name)
             paths = [root_dir]
-            paths.extend([p['path'].rstrip('/')
-                          for p in cfs.ls_dir(root_dir, depth)])
+            with cfs:
+                paths.extend([p['path'].rstrip('/')
+                              for p in cfs.ls_dir(root_dir, depth)])
         except (cephfs.ObjectNotFound, cephfs.PermissionError):
             paths = []
         return {'paths': paths}

</pre>

Reply:
@idryomov  I've updated the commit message to include details of how in the dashboard module the CephFS mount cleanup is done by the cython libcephfs binding's finalizer method, `__dealloc__()`.  I've also created a  dashboard tracker ticket https://tracker.ceph.com/issues/63515. I will post a WIP PR for that dashboard tracker ticket. Hope this is sufficient to complete this cleanup PR.

Reply:
> the CephFS mount cleanup is done by the cython libcephfs binding's finalizer method, `__dealloc__()`

I'm not sure if that `__delalloc()__` implementation is 100% safe, given:

> [...] you should avoid invoking any Python operations which might touch the object. In particular, don’t call any other methods of the object or do anything which might cause the object to be resurrected. It’s best if you stick to just deallocating C data.

and the fact that it _is_ calling `self.shutdown()`.  `shutdown()` doesn't do much but call the C API, but still...

This spans beyond CephFS bindings though, so out of scope here.